precompiles/referenda/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+		// Fetch data from pallet
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let ref_count = ReferendumCount::<Runtime>::get();
+		log::trace!(target: "referendum-precompile", "Referendum count is {:?}", ref_count);
+
+		Ok(ref_count.into())
+	}
+
+	#[precompile::public("submissionDeposit()")]
+	#[precompile::view]
+	fn submission_deposit(handle: &mut impl PrecompileHandle) -> EvmResult<U256> {
+		// Fetch data from pallet
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let submission_deposit = Runtime::SubmissionDeposit::get();
+		log::trace!(target: "referendum-precompile", "Submission deposit is {:?}", submission_deposit);
+
+		Ok(submission_deposit.into())
+	}
+
+	#[precompile::public("decidingCount(uint256)")]
+	#[precompile::view]
+	fn deciding_count(
+		handle: &mut impl PrecompileHandle,
+		track_id: SolidityConvert<U256, u16>,
+	) -> EvmResult<U256> {
+		// Fetch data from pallet
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let track_id: TrackIdOf<Runtime> = track_id
+			.converted()
+			.try_into()
+			.map_err(|_| RevertReason::value_is_too_large("Track id type").into())
+			.in_field("track")?;
+		let deciding_count = DecidingCount::<Runtime>::get(track_id);
+		log::trace!(
+			target: "referendum-precompile", "Track {:?} deciding count is {:?}",
+			track_id,
+			deciding_count
+		);
+
+		Ok(deciding_count.into())
+	}
+
+	#[precompile::public("trackInfo(uint256)")]
+	#[precompile::view]
+	fn track_info(
+		handle: &mut impl PrecompileHandle,
+		track_id: SolidityConvert<U256, u16>,
+	) -> EvmResult<(U256, U256, U256, U256, U256, U256)> {
+		// Fetch data from pallet
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let track_id: TrackIdOf<Runtime> = track_id
+			.converted()
+			.try_into()
+			.map_err(|_| RevertReason::value_is_too_large("Track id type").into())
+			.in_field("track")?;
+		let tracks = Runtime::Tracks::tracks();
+		let index = tracks
+			.binary_search_by_key(&track_id, |x| x.0)
+			.unwrap_or_else(|x| x);
+		let track_info = &tracks[index].1;
+
+		Ok((
+			track_info.max_deciding.into(),
+			track_info.decision_deposit.into(),
+			track_info.prepare_period.into(),
+			track_info.decision_period.into(),
+			track_info.confirm_period.into(),
+			track_info.min_enactment_period.into(),
+		))
+	}
+
+	/// Propose a referendum on a privileged action.
+	///
+	/// Parameters:
+	/// * proposal_origin: The origin from which the proposal should be executed.
+	/// * proposal: The proposed runtime call.
+	/// * at: If true then AT block_number, else AFTER block_number
+	/// * block_number: Inner block number for DispatchTime
+	#[precompile::public("submit(uint8,bytes,bool,uint32)")]
+	fn submit(
+		handle: &mut impl PrecompileHandle,
+		proposal_origin: u8,
+		proposal: BoundedBytes<GetCallDataLimit>,
+		at: bool,
+		block_number: u32,
+	) -> EvmResult {
+		let gov_origin: GovOrigin = proposal_origin.try_into().map_err(|_| {
```

Comment:
Use a new custom function `proposal_origin_from_solidity_u8` to parse form natural indexing instead of TrackId

Reply:
The ReferendaPrecompile is generic over GovOrigin: TryFrom<u8> because we may have different GovOrigin for different runtimes. So the `proposal_origin_from_solidity_u8` would need to be generic as well and there is already TryFrom<u8> for GovOrigin implemented for TrackId.

Reply:
In this case we have to delete the Solidity enum `Origin` and rename the solidity parameter to `trackId`